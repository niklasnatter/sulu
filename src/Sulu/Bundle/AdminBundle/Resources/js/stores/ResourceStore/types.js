// @flow
import type {IObservableValue} from 'mobx/lib/mobx';

export type ObservableOptions = {
    locale?: ?IObservableValue<string>,
};
