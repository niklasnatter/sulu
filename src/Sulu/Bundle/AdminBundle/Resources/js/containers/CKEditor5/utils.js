// @flow
function addLinkConversion(editor: Object, tag: string, internalAttribute: string, tagAttribute: string) {
    // implementation is inspired by the official ckeditor5-link package:
    // https://github.com/ckeditor/ckeditor5/blob/v23.0.0/packages/ckeditor5-link/src/linkediting.js

    editor.model.schema.extend('$text', {allowAttributes: internalAttribute});

    editor.conversion.for('upcast').elementToAttribute({
        view: {
            name: tag,
            attributes: {
                [tagAttribute]: true,
            },
        },
        model: {
            key: internalAttribute,
            value: (viewElement) => viewElement.getAttribute(tagAttribute),
        },
    });

    editor.conversion.for('downcast').attributeToElement({
        model: internalAttribute,
        view: (attributeValue, writer) => {
            return writer.createAttributeElement(tag, {[tagAttribute]: attributeValue});
        },
    });
}

function findModelItemInSelection(editor: Object) {
    const firstPosition = editor.model.document.selection.getFirstPosition();
    return firstPosition.textNode || firstPosition.nodeBefore;
}

function findViewLinkItemInSelection(editor: Object, linkTag: string) {
    const selection = editor.editing.view.document.selection;
    const firstPosition = selection.getFirstPosition();

    return firstPosition.getAncestors().find(
        (ancestor) => ancestor.is('attributeElement') && ancestor.name === linkTag
    );
}

export {addLinkConversion, findModelItemInSelection, findViewLinkItemInSelection};
