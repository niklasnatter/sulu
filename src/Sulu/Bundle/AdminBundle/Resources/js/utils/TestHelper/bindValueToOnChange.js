// @flow
import React from 'react';
import type {Element} from 'react';

const WrapperComponent = (props: {children: Element<*>}) => {
    const component = React.Children.only(props.children);
    const [boundValue, setBoundValue] = React.useState(component.props.value);

    const wrappedOnChange = (newValue, ...remainingParameters) => {
        setBoundValue(newValue);
        component.props.onChange(newValue, ...remainingParameters);
    };

    return React.cloneElement(component, {value: boundValue, onChange: wrappedOnChange});
};

const bindValueToOnChange = (element: Element<*>) => {
    return <WrapperComponent>{element}</WrapperComponent>;
};

// our form components are implemented as controlled components. to test them with @testing-library/react, we
// need to update the "value" that is passed to the controlled component when its "onChange" callback is fired.
// if we dont do this, the component will read the old "value" when multiple events are triggered. for example,
// "userEvent.type()" will trigger an event for each keystroke.
export default bindValueToOnChange;
