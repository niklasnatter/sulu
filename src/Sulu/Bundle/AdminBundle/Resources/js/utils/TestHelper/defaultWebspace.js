// @flow

const defaultWebspace = {
    _permissions: {},
    allLocalizations: [],
    customUrls: [],
    defaultTemplates: {},
    key: 'webspace-key',
    localizations: [],
    name: 'webspace-name',
    navigations: [],
    portalInformation: [],
    resourceLocatorStrategy: {
        inputType: 'resource-locator-strategy-input-type',
    },
    urls: [],
};

export default defaultWebspace;
