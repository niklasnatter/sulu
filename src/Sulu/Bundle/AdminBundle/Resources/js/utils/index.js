// @flow
import {arrayMoveImmutable as arrayMove} from 'array-move';
import {buildQueryString} from './Request';
import {createAjv} from './Ajv';
import {transformBytesToReadableString} from './Bytes';
import {transformDateForUrl} from './Date';
import {translate} from './Translator';
import clipboard from './clipboard';

export {
    arrayMove,
    buildQueryString,
    clipboard,
    createAjv,
    transformBytesToReadableString,
    transformDateForUrl,
    translate,
};
