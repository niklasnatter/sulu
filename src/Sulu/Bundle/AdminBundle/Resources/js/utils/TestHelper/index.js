// @flow
import defaultWebspace from './defaultWebspace';
import listAdapterDefaultProps from './listAdapterDefaultProps';
import fieldTypeDefaultProps from './fieldTypeDefaultProps';
import findWithHighOrderFunction from './findWithHighOrderFunction';

export {
    defaultWebspace,
    listAdapterDefaultProps,
    fieldTypeDefaultProps,
    findWithHighOrderFunction,
};
