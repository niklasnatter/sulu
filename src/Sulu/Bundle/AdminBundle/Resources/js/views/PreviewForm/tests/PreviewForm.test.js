// @flow
import {mount, render} from 'enzyme';
import mockReact from 'react';
import {findWithHighOrderFunction} from '../../../utils/TestHelper';
import ResourceStore from '../../../stores/ResourceStore';
import type {Route} from '../../../services';

const React = mockReact;

jest.mock('../../../stores/ResourceStore', () => jest.fn());

jest.mock('jexl', () => ({
    evalSync: jest.fn().mockImplementation((expression) => {
        if (undefined === expression) {
            throw new Error('Expression cannot be undefined');
        }

        return expression === 'nodeType == 1';
    }),
}));

jest.mock('../../Form', () => class FormMock extends mockReact.Component<*> {
    resourceFormStore = {
        data: {
            testKey: 'test-value',
        },
    };

    render() {
        return <div>form view mock</div>;
    }
});

jest.mock('../../../containers/Sidebar/withSidebar', () => jest.fn((Component) => Component));

beforeEach(() => {
    jest.resetModules();
});

test('Should render Form view', () => {
    const resourceStore = new ResourceStore('snippet', 1);

    const route: Route = ({
        options: {
            previewCondition: 'nodeType == 1',
        },
    }: any);
    const router = ({
        route: route,
    }: any);

    const PreviewForm = (require('../PreviewForm').default: any);

    expect(render(
        <PreviewForm locales={[]} resourceStore={resourceStore} route={route} router={router} />
    )).toMatchSnapshot();
});

test('Should initialize preview sidebar when previewCondition evaluates to true', () => {
    const resourceStore = new ResourceStore('snippet', 1);

    const route: Route = ({
        options: {
            previewCondition: 'nodeType == 1',
        },
    }: any);
    const router = ({
        route: route,
    }: any);

    // require preview form to trigger call of withSidebar mock and retrieve passed function
    const PreviewForm = (require('../PreviewForm').default: any);
    const withSidebar = (require('../../../containers/Sidebar/withSidebar'): any);
    const Form = (require('../../Form'): any);
    const sidebarFunction = findWithHighOrderFunction(withSidebar, Form);

    // mount PreviewForm and call function that was passed to withSidebar
    const previewForm = mount(<PreviewForm locales={[]} resourceStore={resourceStore} route={route} router={router} />);
    const sidebarConfig = sidebarFunction.call(previewForm.instance());

    // check if function that was passed to withSidebar returns the correct SidebarConfig
    expect(sidebarConfig.view).toEqual('sulu_preview.preview');
    expect(sidebarConfig.sizes).toEqual(['medium', 'large']);
    expect(sidebarConfig.props.router).toEqual(router);
    expect(sidebarConfig.props.formStore).toBeDefined();

    // check if evalSync was called with correct parameters during function call
    const jexl = (require('jexl'): any);
    expect(jexl.evalSync).toBeCalledWith( 'nodeType == 1', {testKey: 'test-value'});
});
