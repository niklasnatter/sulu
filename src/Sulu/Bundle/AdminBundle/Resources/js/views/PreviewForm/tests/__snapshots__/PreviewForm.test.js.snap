// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Form view 1`] = `
<div>
  form view mock
</div>
`;
